tools/go.mod

```diff
@@ -5,11 +5,11 @@ go 1.15
 require (
 	github.com/a8m/tree v0.0.0-20201026183218-fce18e2a750e
 	github.com/aws/aws-sdk-go v1.35.5
+	github.com/bufbuild/buf v0.30.1
 	github.com/client9/misspell v0.3.4
 	github.com/elazarl/go-bindata-assetfs v1.0.1
-	github.com/golang/protobuf v1.3.4
```

Comment:
Note that separately, this is a change you're going to want to extensively test - while advertised as compatible, there were major changes, including breaking changes on edge cases, from golang/protobuf 1.3.5 to 1.4.0. Version 1.4.0 was re-implemented in terms of v2 github.com/protocolbuffers/protobuf-go, but while advertised as providing compatibility, this is actually only per their [compatibility guarantee](https://github.com/golang/protobuf#compatibility), which makes exceptions, and there have been numerous documented cases of underlying behavior changing in a way that has broken users (do a search on their github issues).

You'll need to upgrade eventually, so you might as well do it now, and if you're not doing anything overly complicated with Protobuf, you might be fine, just a heads up!

Reply:
@krishicks the spot this worries me with is with external CSI plugins... we don't own the protobuf, we can't extensively test these because there's 100+ of them and many require specific vendor hardware, and we see all kinds of specification compliance issues with them. If we're going to merge this we should either do it and land the fix in a beta for 1.0 or wait until _after_ 1.0-final ships so that we have time to test at least the simple "known good" cloud vendor plugins.

Reply:
I noticed this, but on recompiling the protos found no change. But thanks for the extra info! I _have_ noticed diffs when upgrading e.g. grpc-gateway that gave me enough pause that I decided to roll back the upgrade in other projects.

Reply:
A separate recommendation: I wouldn't track tool dependencies in your `go.mod` at all - I think this is why it can lead to problems. We track them via `make`, for example https://github.com/bufbuild/buf/blob/master/make/go/dep_golangci_lint.mk. This is likely way over-engineered, but the main concept:

```make
FOO_VERSION := v1.2.3
cd /tmp; GOBIN=/path/to/tools/bin GO111MODULE=on go get github.com/foo/foo@$(FOO_VERSION)
```

Doing the `cd /tmp` avoids `go` touching your `go.mod`. Go 1.16 allow this workflow via `go install` https://tip.golang.org/doc/go1.16#go-command

This means that you can have your code dependencies stay however you want (for example, staying on 1.3.4 for golang/protobuf), and install tools that may have upgraded past this themselves (I spent about 10 hours solving all the issues a few months ago, although granted bufbuild/buf has some intricate usage of golang/protobuf).

Reply:
(My above comment was directed at bufdev)

@tgross The upgrade of protobuf here is because of adding `buf` to tools. It wasn't the case that `protoc` was installed via tools before; so if there's concern about this I think we should _not_ install buf via `make deps` and continue to point at the instructions in the README, which were updated to point at buf rather than protoc.

Reply:
> Go 1.16 allow this workflow via go install https://tip.golang.org/doc/go1.16#go-command

We're pinned to go1.15 but once we upgrade to go1.16 that seems like reasonable advice. Not for this PR though.

> @tgross The upgrade of protobuf here is because of adding buf to tools. It wasn't the case that protoc was installed via tools before; so if there's concern about this I think we should not install buf via make deps and continue to point at the instructions in the README, which were updated to point at buf rather than protoc.

Sorry, I might be missing something... shouldn't our README instructions say to use `make deps` so that the installation instructions match exactly what CI will do? And is buf forcing an upgrade to our protobufs version for some reason?

Reply:
Note Go 1.16 isn’t due to come out until February, so using it for this workflow is future-looking advice :-)

Reply:
I've reverted the change that adds `buf` to tools to avoid the upgrade of golang/protobuf.